src/Service/ProducerEventProcessor.cpp

```diff
@@ -0,0 +1,371 @@
+using orbit_grpc_protos::IntrospectionScope;
+using orbit_grpc_protos::ModuleUpdateEvent;
+using orbit_grpc_protos::ProducerCaptureEvent;
+using orbit_grpc_protos::SchedulingSlice;
+using orbit_grpc_protos::ThreadName;
+using orbit_grpc_protos::ThreadStateSlice;
+using orbit_grpc_protos::TracepointEvent;
+
+template <typename T>
+class InternedPool final {
+ public:
+  InternedPool() = default;
+
+  // Return pair of <id, assigned>, where assigned is true if the entry was assigned a new id
+  // and false if returning id for already existing entry.
+  std::pair<uint64_t, bool> GetOrAssignId(const T& entry) {
+    absl::MutexLock lock(&entry_to_id_mutex_);
+    auto it = entry_to_id_.find(entry);
+    if (it != entry_to_id_.end()) {
+      return std::make_pair(it->second, false);
+    }
+
+    uint64_t new_id = id_counter_++;
+    auto [unused_it, inserted] = entry_to_id_.insert_or_assign(entry, new_id);
+    CHECK(inserted);
+    return std::make_pair(new_id, true);
+  }
+
+ private:
+  uint64_t id_counter_ = 1;  // 0 is reserved for invalid_id
+  absl::flat_hash_map<T, uint64_t> entry_to_id_;
+  absl::Mutex entry_to_id_mutex_;
+};
+
+class ProducerEventProcessorImpl : public ProducerEventProcessor {
+ public:
+  ProducerEventProcessorImpl() = delete;
+  explicit ProducerEventProcessorImpl(CaptureEventBuffer* capture_event_buffer)
+      : capture_event_buffer_{capture_event_buffer} {}
+
+  void ProcessEvent(uint64_t producer_id, ProducerCaptureEvent event) override;
+
+ private:
+  void ProcessFullAddressInfo(FullAddressInfo* full_address_info);
+  void ProcessFullCallstackSample(FullCallstackSample* full_callstack_sample);
+  void ProcessFunctionCall(FunctionCall* function_call);
+  void ProcessFullGpuJob(FullGpuJob* full_gpu_job_event);
+  void ProcessGpuQueueSubmission(uint64_t producer_id, GpuQueueSubmission* gpu_queue_submission);
+  void ProcessInternedCallstack(uint64_t producer_id, InternedCallstack* interned_callstack);
+  void ProcessInternedCallstackSample(uint64_t producer_id, CallstackSample* callstack_sample);
+  void ProcessInternedString(uint64_t producer_id, InternedString* interned_string);
+  void ProcessIntrospectionScope(IntrospectionScope* introspection_scope);
+  void ProcessModuleUpdateEvent(ModuleUpdateEvent* module_update_event);
+  void ProcessSchedulingSlice(SchedulingSlice* scheduling_slice);
+  void ProcessThreadName(ThreadName* thread_name);
+  void ProcessThreadStateSlice(ThreadStateSlice* thread_state_slice);
+  void ProcessFullTracepointEvent(FullTracepointEvent* full_tracepoint_event);
+
+  void SendStringAndKeyEvent(uint64_t key, std::string value);
+
+  CaptureEventBuffer* capture_event_buffer_;
+
+  InternedPool<std::vector<uint64_t>> callstack_cache_;
+  InternedPool<std::string> string_cache_;
+  InternedPool<std::pair<std::string, std::string>> tracepoint_cache_;
+
+  // These are mapping InternStrings and InternedCallstacks from producer ids
+  // to client ids:
```

Comment:
Not necessarily here with this comment, but I would add a small explanation somewhere of the fact that we re-intern from different producers to a single "key space" before sending interned strings/callstacks to the client.

Reply:
added above (explains what ProcessIntern* functions do.)